cli/cmd/summary.go

```diff
@@ -0,0 +1,74 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+type syncSummary struct {
+	CliVersion          string `json:"cli_version"`
+	DestinationErrors   uint64 `json:"destination_errors"`
+	DestinationName     string `json:"destination_name"`
+	DestinationPath     string `json:"destination_path"`
+	DestinationVersion  string `json:"destination_version"`
+	DestinationWarnings uint64 `json:"destination_warnings"`
+	Resources           uint64 `json:"resources"`
+	SourceErrors        uint64 `json:"source_errors"`
+	SourceName          string `json:"source_name"`
+	SourcePath          string `json:"source_path"`
+	SourceVersion       string `json:"source_version"`
+	SourceWarnings      uint64 `json:"source_warnings"`
+	SyncID              string `json:"sync_id"`
+}
+
+func persistSummary(filename string, summaries []syncSummary) error {
+	// if filename is not specified then we don't need to persist the summary and can return
+	if filename == "" {
+		return nil
+	}
+	err := checkFilePath(filename)
+	if err != nil {
+		return fmt.Errorf("failed to validate summary file path: %w", err)
+	}
+	for _, summary := range summaries {
```

Comment:
What's the reason to use `jsonl` instead of JSON and dump the whole array instead? Are we expecting consumers to read the output line by line? It's dumped in a single operation anyway and not streamed

Reply:
Appending to JSONL is much simpler because we don't have to deal with competing syncs writing to the same location at the same time or subsequent syncs overwriting the file.

Also because each source <> destination is an entry this makes it much easier to see what succeeded in the case of partial successes

Reply:
> because we don't have to deal with competing syncs writing to the same location at the same time or subsequent syncs overwriting the file

Can you explain what `competing syncs` mean? Do you mean multiple processes of the CLI? Not sure how concurrent writes will work in that case

> subsequent syncs overwriting the file

Don't we want to overwrite the file? Maybe the question to ask it how are we expecting someone to consume the file? If it's always being appended to won't consumers need to handle ignoring information they've already read?

Reply:
> Also because each source <> destination is an entry this makes it much easier to see what succeeded in the case of partial successes

What does partial success mean? Don't we exist early on errors? We could still write the data for the sources that were successful it that's what you mean by partial success, and maybe add additional information on the ones that didn't ran. Otherwise you'd need to compare the spec to the JSON to know which ones were successful

Reply:
We've decided to release this as preview and keep the implementation as is for now until further feedback